### PR TITLE
Improving memory and time efficiency

### PR DESCRIPTION
**Pull request by tbrandirali**
Refactored implementation of prefix tree to support pickling instances and improve memory usage.
Removed visualizer and related dependencies due to incompatibility with latest changes.
Added script to analyze memory allocation for the prefix tree.

### CODE REVIEW

File: whatthelog/prefixtree/sparse_matrix.py [hunk 1]
```diff
@@ -102,17 +124,19 @@ def bisearch(self, arr: List[str], target: str) -> int:
...
+    def get_values(self, index: int) -> Tuple[int, int, str]:
         """
-        Retrieves the value part of the SparseMatrix entry (or the key if key is True).
+        Retrieves the tuple of coordinates and value for the given index.
+        Raises an exception if the index is out of bounds.
+        :param index: the index of the tuple to fetch.
+        :return a tuple in the form(start_node, end_node, edge_value)
         """
-        first: bool = False
+        if index < 0 or index >= len(self.list):
```
Comment (ThomasWerthenbach) [inline whatthelog/prefixtree/sparse_matrix.py]:
```suggestion
        if index < 0 or index >= self.size:
```
Efficiency :)

File: whatthelog/prefixtree/sparse_matrix.py [hunk 2]
```diff
@@ -126,4 +150,7 @@ def get_parents(self, i: int) -> List[int]:
         for item in copy:
             parent, child, props = item.split('.', 2)
             reverse.append(f"{child}.{parent}.{props}")
-        return [tup[0] for tup in self.__binary_search_partial(reverse, i)]
+        return [tup[0] for tup in self.__find_children(reverse, i)]
+
+    def __len__(self):
+        return len(self.list)
```
Comment (ThomasWerthenbach) [inline whatthelog/prefixtree/sparse_matrix.py]: Same as above

Comment (ThomasWerthenbach) [reply whatthelog/prefixtree/sparse_matrix.py]:
```suggestion
        return self.size

```

**Review by ThomasWerthenbach — CHANGES_REQUESTED** [review]

File: whatthelog/prefixtree/prefix_tree_factory.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+                        log + " was not identified as a valid log.")
+
+                exists = False
+
+                for node in nodes:
+                    if template in node.properties.log_templates:
+                        parent = node
+                        nodes = prefix_tree.get_children(parent)
+                        exists = True
+                        continue
```
Comment (gcalin) [inline whatthelog/prefixtree/prefix_tree_factory.py]:
```suggestion
                        break
```

**Pull request merged**